### ISSUE TITLE
FileStreamDisconnectUploadError when checking file validity

### ISSUE THREAD

**Issue by FreddyPoly** [jaydenseric/apollo-upload-server#65]
Hello,

# The problem

I have a problem when checking for the validity of the files sent by multipart request from the client side.

The client sends a multipart request with maximum 5 different files, then the server loops through them to check if every file is an instance of `stream.Readable`. While looping through the files an error occures: `FileStreamDisconnectUploadError: Request disconnected during file upload stream parsing.`.

The first file is checked but then the others files are never read.

This is what my loop looks like:

```
for (const pic of data.pictures) {
      const resolved = await pic;
      if (!(resolved.stream instanceof stream.Readable)) return new Error('Wrong file');
}
```

The error is raised when executing the line `const resolved = await pic;`.
The first file fills the `resolved` variable but it never gets filled again with the next files.

# What I tried

- I tried to destroy the stream after the check: `resolved.stream.destroy();`
- I tried to replace the `for...of` loop with a `map`

# Some thoughts

Before making the requests by the client, I sent the requests using curl according to these [examples](https://github.com/jaydenseric/graphql-multipart-request-spec). But the error still occured.

I work on a Macbook with MacOS Sierra 10.12.6 and curl 7.54.0 which uses `SecureTransport`, as it is the native TLS library.

What is interesting is that my coworkers, which work with the same Macbook as mine but with MacOS High Sierra and curl 7.54.0 which uses `LibreSSL` as TLS library, manage to get the requests work with the server. Of course we tried with the exact same requests and on a server running on my mac.

Do you have any idea why the error occures and why does the `await pic` never manage to complete ?

Thanks for your time.

**Comment by koresar** [comment]
@jaydenseric the `apollo-upload-server` module crashes the whole node.js process if error happens somewhere in the middle. But my code have all the error handlers possible.

```
events.js:183
      throw er; // Unhandled 'error' event
      ^

FileStreamDisconnectUploadError: Request disconnected during file upload stream parsing.
    at IncomingMessage.request.on (/Users/koresar/code/my-project/node_modules/apollo-upload-server/lib/middleware.js:151:15)
    at emitNone (events.js:106:13)
    ...SNIP node.js internals...
```

My code is exactly the same as @FreddyPoly, but without TLS (plain localhost HTTP).

However, that happens occasionally. Could not find stable scenario. Sorry.

**Comment by jaydenseric** [comment]
I think you will run into problems when attempting to `await` in a for loop.

It is better to use `Promise.all`, or the [`promises-all`](https://www.npmjs.com/package/promises-all) approach [used in the example API](https://github.com/jaydenseric/apollo-upload-examples/blob/master/api/resolvers.mjs#L56).

**Comment by koresar** [comment]
@jaydenseric I kind of thought it's vice versa. :) I should await sequentially.

**Comment by jaydenseric** [comment]
@koresar I'm just thinking out loud, and suggestions could be wrong. Node.js streams are a trial and error nightmare.

@FreddyPoly I'm not quite understanding why you are doing `instanceof stream.Readable`. Shouldn't all file upload streams should be of the same type? If the first file fails that check, then does the `new Error` cause the loop, and resolver to bomb? Because if you throw an error and send and early response you would expect all the promises of the remaining uploads to reject with `FileStreamDisconnectUploadError`. You could catch them all, and ignore them if you can somehow signal to the handlers that they were deliberately caused errors.

**Comment by FreddyPoly** [comment]
I'm sorry I don't quite get your point of view when you say:

> You could catch them all, and ignore them if you can somehow signal to the handlers that they were deliberately caused errors.

- Do you mean catching the errors I throw while looping through the files ? So it doesn't reject all the promises ?
- And shouldn't I check for the type of file ? Couldn't I get a file which is not a valid `stream.readable` ? I checked for "security" reasons.

I'm sorry to bother you with what must be a misunderstanding from my part, this is really a great library you made and I'm glad I get to work with it. GraphQL is so easy to work with and adding a system to get files upload is what makes it completely usable.

What I try to do is check if all the files I get from the client are indeed `stream.Readable` and if every file is an image, before doing anything with them.

In my resolver, my first action is to call a validator service to check the files:

```
Mutation: {
    add: async (_, data) => {
      const error = await validationService.filesValidator(data);
      if (error) throw error;

      const offer = await Offer.add(data);

      return offer;
    },
}
```

If the files are not valid I expect my server to throw an error according to the type of error.

My validation service is where I had my `for...of` loop running. Now I try to replace it with `Promise.all` and a `map` as you suggested and as it is stated in the docs.

```
async processFile(file) {
    const resolved = await file;

    if (!(resolved.stream instanceof stream.Readable)) return new Error('Wrong file');
    if (!(/\.(gif|jpg|jpeg|tiff|png)$/i).test(resolved.filename)) return new Error('The picture must be a valid file (gif, jpg, jpeg, tiff, png)');

    return null;
  },
async filesValidator(data) {
    if (!data.pictures) return new Error('An offer needs at least 1 picture to be created');
    if (data.pictures.length > 5) return new Error('A maximum of 5 pictures per offer is allowed');

    await Promise.all(data.pictures.map((pic) => {
      this.processFile(pic.file);

      return pic;
    }));
    // gets executed before the end of the map loop

    return null;
  },
```

The code above "works" but the loop in the `Promise.all` doesn't finish before executing the rest of the code. Adding an `await this.processFile(pic.file)` results in the same error I had formerly.

I'm currently trying to change my implementation but I don't get a stable version. I'll post my solution if I get it to work.

Thanks again for your help and your responses.

**Comment by FreddyPoly** [comment]
Ok it definitely looks like my Macbook has some configuration that makes the request fail. My coworkers who have the same OS version as mine don't manage either to request the server without the error.

So those who are on MacOS High Sierra doesn't have any problem and those who are on MacOS Sierra can't get the request to succeed.

- Maybe the multipart requests aren't build the same way depending on the OS ?
- The file streams aren't handled the same way depending on the OS ?

@jaydenseric Thanks again for your time and your help but I don't think my problem comes from your library or the implementation of the resolver. I guess the OS or some configuration is the key here, pretty hard to know exactly what is the cause. Do you have any idea on what could be different between the two ?

@koresar Did you manage to solve the error ? I would be curious to know how if that's the case and if you don't mind to share your code :)

**Comment by jaydenseric** [comment]
@FreddyPoly Although the OS may be the same, what about the hardware? Is one SSD, more RAM, or better CPU? Maybe it is a race condition somewhere.

**Comment by koresar** [comment]
I couldn't solve it unfortunately. It's not reproducible when I try.

Sad thing it happened in production, but only once.

It all began after I introduced the for-of loop with await in it.

Can it be a missing "error" event handler of the incoming readable stream?
Especially when there are multiple files in the request.

**Comment by FreddyPoly** [comment]
@jaydenseric We changed computers recently and we all have the exact same [Macbook](https://www.apple.com/shop/buy-mac/macbook-pro?product=MPXQ2LL/A&step=config#), so this mustn't be a hardware problem.

I'll investigate some more to find what's the cause of the error.

**Comment by hkjeffchan** [comment]
I ran into similar problem.
Solution: Create a new promise for your operation on the stream and then await Promise.all for your new promises. Do NOT await Promise.all for the streams to resolve and then do your operation with map. The first stream will block all other streams to resolve.

This is similar to RXJS switchMap and then combineAll

**Comment by koresar** [comment]
Trying to understand your suggestion @hkjeffchan. If I rephrase your advice would it sound right?

> If a stream fails you should break the loop and NOT process any further file streams.

**Comment by FreddyPoly** [comment]
I don't get your solution @hkjeffchan . Could you provide an example please ? I can't get this to work.

As I said I investigated to see what was happening and indeed the second file never gets resolved. The first file gets resolved without any problem but the busboy's parser never executes the `on file` event for the next files.

And this happens only if in the resolver, I `await Promise.all` the map to loop through the files I received.

**Comment by FreddyPoly** [comment]
Yes indeed if I want the loop to process all the files I need to consume the stream for each one. Changing my code to:

```
for (const pic of data.pictures) {
  const resolved = await pic.file;

  if (!(resolved.stream instanceof stream.Readable)) return new Error('Wrong file');
  if (!(/\.(gif|jpg|jpeg|tiff|png)$/i).test(resolved.filename)) return new Error('The picture must be a valid file (gif, jpg, jpeg, tiff, png)');

  resolved.stream.resume();
 }
```

Makes the `for...of` loop process al the files in an async way. Just as I wanted it to behave.

However, later I need to get the stream once again to upload the picture to AWS. But this time the stream is empty. I guess that if I consume the stream once it gets empty the next time ?

Can someone confirm this assumption of the empty stream ? I guess I'll have to change my workflow if that's the case.

**Comment by hkjeffchan** [comment]
@FreddyPoly 
You may pipe the stream to a temp file and then upload later. But it is better to pipe the stream to AWS upload if that supports stream.

By the way, you need to `resume()` before returning the new error. Your code will block and raise timeout error in busboy if you pass a wrong file type, right?

**Comment by jaydenseric** [comment]
@koresar I don't think that is the right way to do it, in a loop like that. What if awaiting the S3 upload for the first file takes a long time, meanwhile the second file upload stream finishes unhandled?

Loop though the upload promises, but create a new promise for the processing of each one so you can await them all to complete asynchronous. See https://github.com/jaydenseric/apollo-upload-server/issues/67#issuecomment-382806925.

**Comment by koresar** [comment]
> What if awaiting the S3 upload for the first file takes a long time, meanwhile the second file upload stream finishes unhandled?

As far as I know the second file stream will be buffered by OS.

Here is how multifile upload HTTP request looks like AFAIK:
```
file 1 name, file encoding
.............................
...FILE CONTENTS GOES HERE...
.............................

file 2 name, file encoding
.............................
...FILE CONTENTS GOES HERE...
.............................

file 3 name, file encoding
.............................
...FILE CONTENTS GOES HERE...
.............................
```

Files are provided sequentially. Thus, until the first file is consumed the second file is not provided.

@jaydenseric please tell me where my understanding is wrong. Thank you in advance.

**Comment by koresar** [comment]
I would like to clear some misunderstandings in this sentence:

> Loop though the upload promises, but create a new promise for the processing of each one so you can await them all to complete asynchronously.

**1. "Loop though the upload promises"**
That's what my code is doing.
```js
for (const file of files) {
```

**2. "create a new promise for the processing of each one"**
That's what my code is doing.
```js
await s3.upload({Key: filename, Body: stream}).promise();
```
Yes, it creates a new promise.

**3. "await them all to complete asynchronously"**
I am awaiting for all of them sequentially. Why? Because the second stream won't be available until the first stream is fully consumed. That's coming from [your spec](https://github.com/jaydenseric/apollo-upload-examples/blob/master/app/readme.md). Files are aligned sequentially.
Awaiting for the second stream does not make sense until the first stream is not yet consumed.
Moreover, the second **file name** is not available until the first **stream** is consumed.

-----

Now, lets talk about the code you are suggesting as the correct https://github.com/jaydenseric/apollo-upload-server/issues/67#issuecomment-382806925

As you can see from the code the user uploads two files. First file is small (photo). Second file is big (video). But:
1) User is on a flaky connection.
2) Photo uploads fine, the video upload fails in the middle of the process.

Result: The code **will not save `new Training()` to the database** with the photo. It will throw in `await Promise.all()`. I will end up with a rubbish file in my storage. User experience degrades. We would have to ask them to re-upload photo.

What I want: The `new Training()` is created in the database with the photo. Ignore the video, we would need to upload it later.

**Comment by jaydenseric** [comment]
@koresar In this figure from [the spec](https://github.com/jaydenseric/graphql-multipart-request-spec) intro, you want your process to look like the second “async” scenario:

![Sync vs async GraphQL multipart request middleware](/jaydenseric/graphql-multipart-request-spec/raw/master/sync-vs-async-graphql-multipart-request-middleware.svg?sanitize=true)

If you await the upload to S3 in each loop over the files, there will be no paralel pink and green bits. Not only is it inefficient, but If you files are big enough Node.js won't be able to keep a buffer big enough for incoming uploads while awaiting S3 uploads for previous ones and you will get strange errors.

**Comment by jaydenseric** [comment]
As far as handling errors, there are many ways to go about it. [In the examples](https://github.com/jaydenseric/apollo-upload-examples/blob/master/api/resolvers.mjs#L56) errors with processing one file won't bomb the whole lot, thanks to [`promises-all`](https://www.npmjs.com/package/promises-all).

**Comment by koresar** [comment]
This picture is wrong. The **Async** pink coloured rectangles are misplaced. This is how streaming works in node.js
(I redraw the pink with red)

![image](https://user-images.githubusercontent.com/1622575/39026362-0903cec0-4490-11e8-924e-119387709dcf.png)

Streaming in node.js is done using 1-4KB chunks. Node.js would not buffer the whole file. That's why stream is called, well, "stream".

**Comment by jaydenseric** [comment]
You right that the chart should be improved (the benefits of async would be even more apparent), but the concerns about awaiting earlier pink bits before processing incoming green bits still stand. What if an upload to cloud takes twice as long as the incoming upload stream?

**Comment by koresar** [comment]
> What if an upload to cloud takes twice as long as the incoming upload stream?

It would behave same way as multipart (or even single file) upload behaved till today. This is long solved problem.
* The node.js (OS level TCP socket) will not be replying "ACK" per each TCP packet as fast. That's it.
* For user it will look like slower upload.

And we, the developers, can do nothing about it. The `Promise.all()` would not solve slow hardware (HDD) or infrastructure (S3) problem.

**Comment by hkjeffchan** [comment]
I just hack the middleware.js from this package to log as much event as possible. After testing with different filesize and sequence samples, can confirm @koresar is correct.

1) files are provided sequentially via multifile upload HTTP request.
2) a for-loop is enough, promise all is not required
3) the root problem of the original code below is the stream is not CONSUMED, not related to the promise all or for-loop

```
for (const pic of data.pictures) {
      const resolved = await pic;
      if (!(resolved.stream instanceof stream.Readable)) return new Error('Wrong file');
}
```

TL:DR
You must consume the stream whenever the file is resolved `const file = await fileList;`, just pipe it to another stream or call `file.stream.resume()` to discard it. Otherwise, the underlying busboy will wait for you to consume the stream chuck and eventually timeout and crash the node server.

**Comment by jaydenseric** [comment]
@koresar I just updated the diagram: https://github.com/jaydenseric/graphql-multipart-request-spec/blob/master/sync-vs-async-graphql-multipart-request-middleware.svg

Here is a diagram illustrating why I don't think you should use `await` on the S3 uploads in a loop:

![multipart-graphql-request-async-issue](https://user-images.githubusercontent.com/1754873/39038168-270767de-44c6-11e8-970d-cf1911400b76.png)

**Comment by koresar** [comment]
The image still a little misleading.

All infrastructures are always much faster than client (browser) network throughput. Otherwise few people uploading files would hung the sever.

In other words file storages like S3 are always faster than user's internet connection.

Also, large file upload is always goes neck to neck with storage stream. Your new image implies that node.js would buffer the entire video file! That's is not how it works in streams.

Streams were invented to make sure large file buffering doesn't happen. Your new image is wrong in both graphs.

**Comment by FreddyPoly** [comment]
So I tried the solution of @hkjeffchan and indeed consuming the stream in the loop makes the process to work. However I need to get the stream once again to upload it to S3, I can't do it at the same time as the validation of the files. I'll try to pipe it to another stream as you suggested or change my workflow.

I close this issue as it isn't a bug but rather a bad use of the library.

Thanks for your help and your time everyone.

**Comment by Access2emma** [comment]
Please am having the same error. Am uploading multiple files and getting this same error. Can you post a working snippet for saving it to local drive. Not S3 in this case?

**Comment by twelve17** [comment]
@jaydenseric : per @koresar 's comments, shouldn't the "Tips" section be updated to not recommend "Promise.all" anymore?  Am I understanding correctly that given that the nature of multipart uploads is sequential, then beginning a second upload in parallel doesn't offer any improvement over streaming them sequentially?

**Comment by mike-marcacci** [comment]
> Am I understanding correctly that given that the nature of multipart uploads is sequential, then beginning a second upload in parallel doesn't offer any improvement over streaming them sequentially?

The HTTP request stream is sequential indeed, but files are not necessarily sent in the same order as you expect (the same GraphQL variable can be used in multiple places, and in different orders). The `graphql-upload` library handles all these scenarios and gives you an API for writing resolvers that lets you focus on your business logic.

You should simply use the standard concurrency patterns of JS to express the data dependancies, and the library will deliver your data in the most efficient way possible given a request. If one upload _depends_ on another having already been processed, use something like sequential `await` expression to describe this; if two uploads _can_ be consumed concurrently, use the language features like `Promise.all` to make this possible, and you will see a performance gain in certain scenarios.

**Comment by koresar** [comment]
Hey Mark.
I was trying to imagine at least one scenario where Promise.all would give
a perform benefit. But could come up with scenarios where it slows things
down or makes server more endangered.

Could you please come up with a scenario where Promise.all would give a
gain (not a loss)?

Thank you

**Comment by mike-marcacci** [comment]
Sure. Let's say you write your resolver so that it processes uploaded image `A`, then image `B`, then image `C`. Your resolver applies a streaming transformation to these images and saves them in S3/Cloud Files/whatever in a `for of` loop, using `await`.

Under **the best conditions**, you end up waiting on the transformation and saving of `A` before beginning the transformation and saving of `B`, which is a total waste since this could be done concurrently. It's possible that for your workload the upload is the slowest part "most of the time" and that this doesn't matter to you in practice, but there is no opportunity for it to benefit from concurrency should the network be particularly fast, or your transform to be particularly slow.

Under **less ideal** conditions, the actually HTTP request (which _you_ don't control), the files may be sent in the order `C`, then `B`, then `A`. Now, your app waits for _all three_ images to be buffered into the filesystem. Then, once `A` finally resolves, you have to wait for its transformation and saving before starting `B`. This is a *huge* waste here.

Finally, we have ways to express concurrency requirements in javascript, and you should use them. If I came across the code I described above, I might assume that there is a side effect to the transformation/saving processes that _requires_ these to be processed sequentially... why else would the author choose such a pattern over a well-established, generally more performant one? Imagine a future developer tasked with adding a high-cost transformation to the above code (say some sort of image compression). She isn't certain _why_ the original author forced sequential execution, so she just adds the transformation to the existing flow control. The change makes the transformation step slower than the upload step in many cases, and now the original assumptions that allowed the sequential flow to perform comparably to the more expressive alternative are no longer true.

All that said, using a sequential flow here will still work correctly, and for many workloads this won't make a noticeable difference. But it is _NOT_ the recommended pattern.

**Comment by koresar** [comment]
Okay. Here is my take on it. I would **not** recommend `Promise.all` because of several reasons.

* Your "less ideal" conditions is impossible. There is literally no scenarios in the world where uploading file B depends on the contents of the file A. (Btw, the great and mighty `graphql-upload` module has the feature called "variables" to cary any data besides the files.)
* The `Promise.all` was invented to execute several things at the same time, like saving data to 3 databases **simultaneously**. Suggesting `Promise.all` implies that we are consuming A, B, and C simultaneously. But this is very false assumption. We are confusing newbie developers with it too much.

I'd recommend removing the `Promise.all` from any code examples as quite misleading. Instead, use this **very readable** code:

```js
async function resolver(root, { files }, context) {
  try {
    for (const file of files) {
      const { stream, filename /*, mimetype, encoding */ } = await file;

      s3.upload({Key: filename, Body: stream}).promise() // <- NOTE! There is no `await`!
        .catch(err => console.error(`Couldn't process ${filename}, skipping it.`));
    }

    return {success: true};
  } catch (err) {
    log.error(err);
    return {success: false, message: err.message};
  }
}
```

By removing the `await` and adding `.catch()` in the marked line I get the same result as using the `Promise.all` described one comment above by @mike-marcacci

**Comment by mike-marcacci** [comment]
> Your "less ideal" conditions is impossible. There is literally no scenarios in the world where uploading file B depends on the contents of the file A. (Btw, the great and mighty graphql-upload module has the feature called "variables" to cary any data besides the files.)

You may want to reread my post... my "less ideal" scenario has nothing to do with dependancies of files on each other, _or_ other data. Instead, it has to do with the HTTP request, which is a linear stream of bytes that can contain uploads in a different order than you expect in your resolver. Consider this example query:

```graphql
query {
  foo: findPeople(byFacePhotos: [$uploadB, $uploadX, $uploadA]) {
    name
  }
  bar: findPeople(byFacePhotos: [$uploadA, $uploadB, $uploadC]) {
    name
  }
}
```

This kind of request is valid GraphQL and _must_ still be valid when using the the upload spec to do variable replacement. Therefore, the upload spec _cannot_ constrain the order of uploads in the HTTP request such that it matches the order of appearance in every list.

It is completely possible that your upload stream will contain the contents of `$uploadC` first, followed by `$uploadB`, then `$uploadA`. You don't get to decide this, because you aren't sending the request; the client is. If your code is waiting on `$uploadA` before processing `$uploadB`, then it is inefficient.

Also, there _are_ scenarios where business logic requires one upload to be present before another becomes useful; just because you haven't run into one doesn't mean they don't exist. Your tone here is painfully hyperbolic and very much in contrast to the meticulous attention to edge cases that has made this library suitable for such a wide range of use cases. If you have good feedback, we want to hear it, and dismissive hand-waving is counterproductive to meaningful communication of an intricate subject.

---

> The Promise.all was invented to execute several things at the same time, like saving data to 3 databases simultaneously. Suggesting Promise.all implies that we are consuming A, B, and C simultaneously. But this is very false assumption. We are confusing newbie developers with it too much.

That's not how promises work. Promises are executed _eagerly_, meaning that they run _as soon as they are created_. Something like `Promise.all` does **not** _cause them to run_ but instead "pauses" the current flow when using `await` (or waits to call any methods registered with `.then()`) until all its promises (which are already running) are resolved.

---

I suspect your example does not work the way you think it works, and if it does, I don't believe it represents a common use case. In my experience, it's been standard practice to ensure the file is stored without error _before_ returning a "successful" response.

Keeping in mind that the HTTP response _won't make it to the browser_ until the request has finished, your resolver should almost certainly be refactored to:

```js
async function resolver(root, { files }, context) {
  try {
    await Promise.all(files.map(async (file) => {
      const { stream, filename /*, mimetype, encoding */ } = await file;
      await s3.upload({Key: filename, Body: stream}).promise();
    }))
    return {success: true};
  } catch (err) {
    log.error(err);
    return {success: false, message: err.message};
  }
}
```

**Comment by koresar** [comment]
Apologies for my unconstructive tone.

I finally see what you mean now. The query example explains it all.

Again, I realised that talking about code but not seeing the code often leads to an unconstructive discussion.

PS: You've refactored the code back it it's original state little above in this thread: https://github.com/jaydenseric/graphql-upload/issues/65#issuecomment-382922064

**Comment by twelve17** [comment]
@koresar @mike-marcacci Thanks for the additional discussion an explanation.  For posterity, I was not so much concerned with the order of file uploads being non-determinant, I was more worried whether the uploads to the second endpoint (s3 in the examples above) would all be initiated in the beginning, meaning whatever upload was _not_ the first one would leave that endpoint's server hanging while the first upload completed, potentially leading to timeouts.  I [created a test repo](https://github.com/twelve17/test-graphql-upload) that performed all these steps, and looking through the log messages I added, it looks like indeed, each promise of the upload to the second endpoint is not initiated until the upload is complete from the originating client.
